share/eds/utils.go

```diff
@@ -46,3 +55,98 @@ func closeAndLog(name string, closer io.Closer) {
 		log.Warnw("closing "+name, "err", err)
 	}
 }
+
+// RetrieveNamespaceFromStore gets all EDS shares in the given namespace from
+// the EDS store through the corresponding CAR-level blockstore. It is extracted
+// from the store getter to make it available for reuse in the shrexnd client.
+func RetrieveNamespaceFromStore(
+	ctx context.Context,
+	store *Store,
+	dah *share.Root,
+	namespace share.Namespace,
+) (shares share.NamespacedShares, err error) {
+	if err = namespace.ValidateForData(); err != nil {
+		return nil, err
+	}
+
+	bs, err := store.CARBlockstore(ctx, dah.Hash())
+	if errors.Is(err, ErrNotFound) {
+		// convert error to satisfy getter interface contract
+		err = share.ErrNotFound
+	}
+	if err != nil {
+		return nil, fmt.Errorf("eds/utils: failed to retrieve blockstore from eds store: %w", err)
+	}
+	defer func() {
+		if err := bs.Close(); err != nil {
+			log.Warnw("closing blockstore", "err", err)
+		}
+	}()
+
+	// wrap the read-only CAR blockstore in a getter
+	blockGetter := NewBlockGetter(bs)
+	shares, err = CollectSharesByNamespace(ctx, blockGetter, dah, namespace)
+	if errors.Is(err, ipld.ErrNodeNotFound) {
+		// IPLD node not found after the index pointed to this shard and the CAR
+		// blockstore has been opened successfully is a strong indicator of
+		// corruption. We remove the block on bridges and fulls and return
+		// share.ErrNotFound to ensure the data is retrieved by the next getter.
+		// Note that this recovery is manual and will only be restored by an RPC
+		// call to SharesAvailable that fetches the same datahash that was
+		// removed.
+		err = store.Remove(ctx, dah.Hash())
+		if err != nil {
+			log.Errorf("eds/utils: failed to remove CAR from store after detected corruption: %w", err)
```

Comment:
Optional/nit: "eds/utils:" feels kinda too generic and useless. Error should be wrapped by the caller.